components/server/src/user/user-controller.ts

```diff
@@ -281,6 +287,64 @@ export class UserController {
                 res.sendStatus(200);
             },
         );
+
+        router.post(
+            "/auth/workspacePageClose/:instanceID",
+            async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+                if (!req.isAuthenticated() || !User.is(req.user)) {
+                    res.sendStatus(401);
+                    log.warn("unauthenticated workspacePageClose", { instanceId: req.params.instanceID });
+                    return;
+                }
+
+                const user = req.user as User;
+                if (user.blocked) {
+                    res.sendStatus(403);
+                    log.warn("blocked user attempted to workspacePageClose", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+
+                const instanceID = req.params.instanceID;
+                if (!instanceID) {
+                    res.sendStatus(400);
+                    log.warn("attempted to workspacePageClose without instance ID", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+                const sessionId = req.body.sessionId;
+                const server = this.createGitpodServer(user);
+                try {
+                    await server.sendHeartBeat({}, { wasClosed: true, instanceId: instanceID });
+                    /** no await */ server
+                        .trackEvent(
+                            {},
+                            {
+                                event: "ide_close_signal",
+                                properties: {
+                                    sessionId,
+                                    instanceId: instanceID,
+                                    clientKind: "supervisor-frontend",
+                                },
+                            },
+                        )
+                        .catch((err) =>
+                            log.warn({ userId: user.id }, "workspacePageClose: failed to track ide close signal", err),
+                        );
+                    res.sendStatus(200);
+                } catch (e) {
+                    log.error("workspacePageClose failed", e);
+                    res.sendStatus(500);
```

Comment:
@mustard-mh I just realised that it is wrong, i.e. if a user tries to access a workspace to which they don't have access it will pop up in our metrics like 500 (i.e. our failure), than it should be 403. Could you have a look sendHeartbeat method and translate exceptions properly? 🙏 cc @gitpod-io/engineering-webapp